### ISSUE TITLE
Non-linear editor with multiple files

### ISSUE THREAD

**Issue by mifi** [mifi/lossless-cut#976]
An often requested feature that would allow a more dynamic timeline, kind of like a traditional editor with multiple files. This is a very big feature, and I'm not sure if this is the direction LosslessCut should go, as many people like the current way of editing. But it's nice to be able to discuss it nonetheless. I think it would go something like this:

- Allow drag-dropping multiple files on the timeline
- Starts out with one big segment per opened file on the timeline, then the user can chop up, duplicate and move segments around like in a traditional editor
- Each segment would be connected to a file (or some other dynamic frame source, like say a fill color)
- Obviously all files still would need the same codec parameters for all their tracks
- Allows add pause between segments #255
- Allows (lossy) adding images and other videos as segments (they would need to be encoded with the same codec)
- This would effectively make the current "Merge/concatenate files" dialog obsolete
- This kind of clashes a bit with the tracks editor (it would need to work per-file)

**Comment by 14v** [comment]
Is it possible to implement duplicating segments first? This may solve many video editing issues and does not need file re-encode.

**Comment by mifi** [comment]
duplicating segments would require an NLE already implemented. But you can still duplicate segments as a workaround by first exporting the segment and duplicating the resulting file, then losslessly merge/concat the files

**Comment by worldpoop** [comment]
Here's how you do it without betraying its essence of Lossless Cut or getting into a quagmire of budding NLE design.  Lossloss cut just needs to allow two more things.  Surely ffmpeg and some piping can handle this:

1.  Setting up a batch movie file merge, on each imported movie you have the option of setting a single set of ying-yangable in and out points.

2.  You have the ability import the same file more than once to the batch, appearing in the batch as a separate instance.

Done.  Now with two simple changes / additions, you have the ability to do things like this:  I import video A and exclusively mark out a middle section for omission.  Then I import and mark video B that will be batch-merged to A.  Then after this B import, I import video A again and mark an inclusive in/out near the beginning that also happens to somewhat overlap the selection made in the first instance in A.  And those are my batch instructions: Two parts of A, all of B, a different part of A.  No timelines, just the same "right-side" vertical batch list -- dear ffmpeg, please merge this file portion to this file portion to this file portion.

Batch export this.  Done.  No multiple "right-side" exports.  No re-imports.  Ecstatic users!

**Comment by FrisoDB** [comment]
This would be a great feature.
When I first used the concatenate tool/batch list I assumed it worked like @worldpoop described, but it did not.

**Comment by tomasz-lasko** [comment]
Hi, in the beginning of this issue you wrote a lot of complicated functionality, which is indeed nonlinear editor. However most of the questions in other issues, which you redirect here, as well as my personal questions is very simple and one could say linear:

I have a batch file list (A, B, C), each file has it's own cutting points set (A1-A2, B1-B2, C1-C2). So now when I want to merge/concatenate (assuming all required conditions like same codec etc etc), **the only new functionality would be so that the concatenate using the selected cuts (output = A1-A2 + B1-B2 + C1-C2)** - apart from simple merging whole input files as it is done now (output = A + B + C)

In conclusion, my opinion is that:
1. Honestly this doesn't seem complicated at all, I hope the libraries you use allow to easily implement that. 
2. This should be a seperate issue (like #423), implemented first. And the NLE in current issue should be a long term far goal.

**Comment by mifi** [comment]
While your use case mighy seem simple at a glance, it is just one of many different use cases that are all **covered by implementing NLE** functionality. There isn’t any library in losslesscut to do NLE with ffmpeg. It’s all done manually and it’s not trivial to add such use cases.

**Comment by tomasz-lasko** [comment]
> it is just one of many different use cases that are all covered by implementing NLE functionality

Thank you. Ok. So my question:    if we first **forget** about all other use cases, and focus on the simplest cases, then what do you think - wouldn't these few cases be most needed, and all the other cases clearly less urgent? 

If no then my suspicion was wrong. But if yes, then I suggest totally splitting these two sets of cases - in short term the simple and important ones, and in some long term all other cases that fall under NLE.

**Comment by mifi** [comment]
As as I see it, the amount of work needed to implement a «simple nle case» is not insignificant. Especially considering how complex the current implementation already is (multi tracks, smart cut, cut/merge and a multitude of options).

**Comment by worldpoop** [comment]
I agree, the objections are perceiving the asks in overly complicated fashion and kind of getting stuck there.  All that's being asked for is functionality that LC can already do, but with the decision list being better organized.  That's it. No libraries are needed to assign sets programmatically to a better organized decision list:  essentially the cuts -- output either to temporarily or intermediary ("keep") files, or "|" , "<" -- then merged.  There is no change to the existing rudimentary timeline functionality represented in the LC GUI.  Nothing needs to be added there.

LC already creates the decision list as it is now in the form of sets listed in a column representing the eventual outputted segments.  The subsequent merge order of the cut sets that LC now lists in a column need not be represented in the GUI again!  It's one nested step instead of having to end the program and restart it.    LC programmatically remains aware of the sets instead of only the user, that's all.  

And it should only goes in one direction start to exit: either a merge list followed by cut list, end; or cut list followed by merge list, end.  (No cut A,B,C, then merge BAC but cutting A again for BA1CA2 or whatever -- that's an NLE.  No, only cut A,B,C, merge BAC.  In other words, it's NOT non-linear, certainly not a timeline based NLE.). One could practically do this with scripts and ffmpeg, except that in LC you have already done the hard work of representing a rudimentary timeline GUI and a decision list of operations to be performed.  Just use them, but with a better (slightly more robustly) organized decision tree.  You are so close already!

**Comment by worldpoop** [comment]
tldr: 

Two text columns, two runs (with no more than two runs possible): cut list (run); merge order list of the cuts (run); end.

Note, LC already does one OR the other.  But it could programmatically feed one TO the other.  That's it.  It's all in the coding, handling of the sets, no libraries needed.

Current LC process (v.1): Cut list (mode) OR merge list (mode), run. End.

Enhanced LC process (v.2):  Cuts list, merge order, run. End.

Enhanced LC process (v.3):  Cuts list, merge order, run (mode) OR merge list, cut list, run (mode). End.

On v.3, you would need to handle some math, so work on v.2. (which your users will LOVE you for).  Assess later if v.3 is worth it -- don't worry about it now.  Just make v.2, that's quick and gorgeous!

**Comment by mifi** [comment]
This sounds more like an extension of #89 `mass export button - I think the first version should just have a function to export all files in the batch list using current settings (without showing the export dialog)`

**Comment by KamilSucharski** [comment]
I needed this feature so much that in the end I created my own lossless video editor, that has this kind of workflow - [Ume Movie Editor](https://sengami.itch.io/ume-movie-editor). It handles multiple videos in one project, has the ability to create subtitles and chapters, has a multi-track UI, and Windows and Android versions. I know it's not be the best place to talk about it, since it's unrelated to LosslessCut, but seeing how many people end up mentioned in this github issue, I think they deserve to know that there is an alternative. I urge everyone to NOT respond to this post, as to not derail/spam this discussion.

<details>
<summary>Click to expand screenshots</summary>

![Image](https://github.com/user-attachments/assets/a013e877-1302-4ad6-b29e-ab3452cbc813)

![Image](https://github.com/user-attachments/assets/7349b6cd-2533-4a0e-93ad-63a9323d0643)

![Image](https://github.com/user-attachments/assets/fde93913-c6c9-48aa-bc49-bd3b0830a04f)

![Image](https://github.com/user-attachments/assets/a166b9c0-eae5-437c-a917-ccbb2faf9419)

![Image](https://github.com/user-attachments/assets/f88c7408-1453-4af8-ac47-1ca77937eb25)

![Image](https://github.com/user-attachments/assets/f8eea027-6b85-475b-9b72-b3f6325344d4)

</details>

**Comment by mifi** [comment]
Cool app, but since it's a closed source paid-only app, I I'm going to mark it as off topic because IMO it feels like an advertisement and I don't think it belongs here. If it was open source, I'd be open to even mentioning it in the README.
